transaction-logger/src/main.rs

```diff
@@ -0,0 +1,731 @@
+                    }
+                    if sender
+                        .send((info, with_addresses, summary.special_events))
+                        .await
+                        .is_err()
+                    {
+                        log::warn!(
+                            "The database connection has been closed. Terminating node queries."
+                        );
+                        return Ok(());
+                    }
+                    *height = height.next();
+                }
+                Err(_) => {
+                    success = false;
+                    // if we failed one height we drop the rest since we likely failed those as
+                    // well. In any case we must do things in order.
+                    check_node_and_wait(&mut node).await?;
+                    break;
+                }
+            }
+        }
+        if success {
+            parallel = std::cmp::min(parallel + 1, max_parallel.into());
+        } else {
+            parallel = 1;
+        }
+    }
+    Ok(())
+}
+
+/// Try to reconnect to the database with exponential backoff, at most
+/// MAX_RECONNECT_ATTEMPTS times.
+async fn try_reconnect(
+    config: &postgres::Config,
+    stop_flag: &AtomicBool,
+) -> anyhow::Result<DatabaseClient> {
+    let mut i = 1;
+    while !stop_flag.load(Ordering::Acquire) {
+        match DatabaseClient::create(config.clone(), postgres::NoTls).await {
+            Ok(c) => return Ok(c),
+            Err(e) if i < MAX_CONNECT_ATTEMPTS => {
+                let delay = std::time::Duration::from_millis(500 * (1 << i));
+                log::error!(
+                    "Could not connect to the database due to {}. Reconnecting in {}ms",
+                    e,
+                    delay.as_millis()
+                );
+                // wait for 2^(i-1) seconds before attempting to reconnect.
+                tokio::time::sleep(delay).await;
+                i += 1;
+            }
+            Err(e) => {
+                log::error!(
+                    "Could not connect to the database in {} attempts. Last attempt failed with \
+                     reason {}.",
+                    MAX_CONNECT_ATTEMPTS,
+                    e
+                );
+                return Err(e.into());
+            }
+        }
+    }
+    anyhow::bail!("The node was requested to stop.")
+}
+
+async fn write_to_db(
+    config: postgres::Config,
+    height_sender: tokio::sync::oneshot::Sender<AbsoluteBlockHeight>, // start height
+    mut receiver: tokio::sync::mpsc::Receiver<TransactionLogData>,
+    stop_flag: Arc<AtomicBool>,
+) -> anyhow::Result<()> {
+    let mut db = try_reconnect(&config, &stop_flag).await?;
+    create_tables(&db).await?;
+
+    let height = get_last_block_height(&db)
+        .await?
+        .map_or(0.into(), |h| h.next());
+    height_sender
+        .send(height)
+        .map_err(|_| anyhow::anyhow!("Cannot send height to the node worker."))?;
+    let mut retry = None;
```

Comment:
I am not entirely sure what you are suggesting. The receiver is this https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html which is not an iterator. The only way to get elements out is via `recv` methods.

The reason for this retry is that if after we retrieve the element from the channel, we then fail to use it, we need to try again, and we can't put things back into the channel (since we only have the receiver here).

I agree it is a bit convoluted, but I'm not sure there's a simpler way.